### ISSUE TITLE
Implement batching and summarization to unit tests

### ISSUE THREAD

**Issue by Witiko** [Witiko/markdown#245]
#255 hotfixes the fact that an increasing number of unit tests runs slowly in continuous integration but does so at the expense of testing all the different TeX formats in pull requests. In addition to speed issues, the unit testing framework also fails fast rather than finishing all tests and showing success / failure ratios.

To improve the situation, we should revert #255 and:

1. Batch unit tests, which would process batches of unit tests, amortizing the start-up time of TeX.
2. Collect statistics and display succeesful and failed tests at the end instead of failing fast.

To make implementation and maintenance easier, we should likely rewrite the framework from [Bash][1] to Python.

 [1]: https://github.com/Witiko/markdown/blob/8634db6e924f8e9de099b5ff3885a0fc64062449/tests/test.sh

**Comment by Witiko** [comment]
Not a priority with #255. Removing the 2.23.0 milestone.

**Comment by Witiko** [comment]
# Tasks

- [ ] Rewrite `test.sh` to `test.sh`, `test.py` and `requirements.txt`
    - Install python3 and python3-virtualenv in Dockerfile.
    - Make `test.sh` produce a virtualenv and install requirements before running `test.py "$@"`.
    - Warm up the cache with the first test before running the remaining tests in parallel.
    - Show progress and ETA with tqdm.
    - When a test fails, report the name of the test file, the exit code, and the expected/actual diff.
- [ ] Implement batching to `test.py`
    - Construct a temporary batch test file with multiple tests, separated in TeX logs.
    - In case of a soft failure (all tests completed successfully), report the failing test.
    - In case of a hard failure (a test crashed TeX), report the expected cause, and bisect the batch.
       When a batch of size 1 fails, report the failing test.
- [ ] Revert https://github.com/lostenderman/markdown/commit/d81ad59b6b13af13a84c7acb08bc4622cafd67bf and fix regressions reported in https://github.com/lostenderman/markdown/pull/153#issuecomment-1627797343.
